### ISSUE TITLE
dependencies.condition is not behaving correctly in subcharts

### ISSUE THREAD

**Issue by gintautassulskus** [helm/helm#4490]
The [example repository](https://github.com/gintautassulskus/chart1) has the following chart structure:

```
- Chart1 depends on
    - Chart2 depends on
        - Postgresql
           condition: postgresql.enabled
```

At the top, we have Chart1 that depends on Chart2. Chart2, in turn, depends on Postgresql chart.
The latter has a condition `postgresql.enabled`.

I would like to override the setting at the Chart1 level. The expected expression does not work:
```
cahrt2:
  postgresql:
    enabled: false
``` 

What does work though is the direct reference to the value:
```
postgresql:
  enabled: true
```

I believe this is an incorrect behaviour, we should refer to the values within dependency charts via chart names.

PS. In the Chart1 directory, `deploy.sh` has the necessary commands to deploy the chart.

Output of `helm version`:
Client: &version.Version{SemVer:"v2.10.0-rc.3", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.10.0-rc.3", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}

**Comment by jascott1** [comment]
Hi @gintautassulskus 

Sorry you are having trouble. First thing, you have a typo in your example values file in chart1 for the `chart2:` key
https://github.com/gintautassulskus/chart1/blob/master/values.yaml#L1

Conditions are evaluated from root of the values (for good or ill thats how it is). You can accomplish what (i think) you want by changing chart2's requirements' postgresql condition to the following:

`condition: chart2.postgresql.enabled,postgresql.enabled`

If you install from chart1, the first condition will "win" and control the installation of postgresql.
If you install from chart2, there will be no chart1 values so the second condition will "win" and control the installation of postgresql.

**Comment by gintautassulskus** [comment]
Thanks, @jascott1! Your solution did the trick.

I am sure this has been discussed before, but will express my concern regarding the behaviour of the `condition` option. The vast majority of charts use something like:

```condition: postgresql.enabled```

Let's suppose I have a `rootChart` that depends on [artifactory](https://github.com/helm/charts/tree/master/stable/artifactory) and [concourse](https://github.com/helm/charts/tree/master/stable/concourse) charts. 
Both of these charts have `condition: postgresql.enabled`.

The only way to configure postgresql within the artifactory and concourse charts at the `rootChart` level is the entry below, which will affect both postgresqls. There is no way to individually configure each of the postgresql. 

```
postgresql:
  enabled: true
```

What's more, your solution will not work with alias, because chart's name is hardcoded.

Finally, the behaviour is also not consistent with the variables used in other parts of the chart.

This sounds like a usability bug to me.

**Comment by blindworld** [comment]
I ran into this same issue with nginx.  I set a conditional requirement to "ingress.enabled" in the requirements.yaml for both parent and children charts.  I was expecting the child's requirements.yaml file to include or exclude dependencies based on the child's value's file, not the parent's values file.  Since the conditional was the same across parent and children, I ended up with install errors due to the duplication.

Thanks to this thread I have a work around, but at the very least it would be nice to note that it behaves differently than other variables in the documentation.

**Comment by grantbachman** [comment]
I just wanted to chime in that this bit me as well. The behavior isn't at all intuitive when considering how subchart variables are referenced from a parent chart's`values.yaml`. There should at least be a note in the docs mentioning this. I can submit a PR if needed!

**Comment by victornoel** [comment]
I'm hitting this bug but in a different way, I have this situation:
- chart1 (this is an umbrella chart I made)
  - chart2 (this is a reusable chart I made)
    - cert-manager
       - webhook

cert-manager has a dependency on webhook and is using `condition: webhook.enabled` for it.

I haven't change the value of this, but the `webhook` chart is installed anyway.

I tried setting in chart1/values.yaml `webhook.enabled: false` but it doesn't change a thing.
I tried with the command line `--set` option, I tried to prefix it with chart1, with chart2, with both, with cert-manager, etc, and none of those work: the chart is enabled.

**Comment by jascott1** [comment]
Hi @victornoel Can you try `chart2.cert-manager.webhook.enabled: false` ?

**Comment by victornoel** [comment]
@jascott1 I've tried all the combinations :) it is as if the fact that the condition is for a sub sub chart makes it to be ignored. If i go into `chart2` and run helm from there, it's ok, without doing anything special. It is easy to reproduce I believe:
- create an empty charts for chart1
- create an empty chart for chart2 physically in the charts directory of chart1
- add cert-manager 0.5.2 in chart2 requirements file

Run helm template or helm install --dry-run --debug for chart1: webhook is enabled while it shouldn't. Go into chart2, run the same: webhook is disabled as it should.

**Comment by jascott1** [comment]
@victornoel did you also add a requirements file under chart1 that includes chart2?

**Comment by victornoel** [comment]
@jascott1 no, should I? I thought the requirements.yaml file was there to manage external versioned charts (downloaded from the web), not those physically available in the charts directory.
If I should add it, what would be the URL to use?

**Comment by rcoh** [comment]
FYI, the hackaround to this alluded to in the question (but I want to call out it for others explicitly) is to include the condition in your root chart values _exactly_ as it appears in the subchart requirements.

Eg. instead of 
```
patroni:
  etcd:
    deployChart: true
```
do:
```
etcd:
  deployChart: true
```

**Comment by kfox1111** [comment]
I just bumped into this issue too. Forcing a subchart to become aware of all the names the parent could embed it as, is not a scalable solution. I think this is clearly a bug.

**Comment by bacongobbler** [comment]
Helm's dependency management solution only handles subchart dependencies one level deep. There is the chart to be installed, and then there are subcharts; there's no middleground. All subcharts are all treated at the same level, so there's no notion of "subchart of subchart" handling with regards to value processing.

This was intended by design to prevent the same subchart being included twice in the same chart. In other words, it was designed to prevent chart developers from installing postgres twice in the same release. Instead, it was designed for the chart to share all of the dependencies together, such that if subchart A and subchart B both rely on postgres, they can rely on the same instance rather than installing two separate instances (and therefore double the compute power).

So in other words, the behaviour of having requirements treated at the root level i.e.

```
postgresql:
  enabled: true
```

Was an intended design choice of Helm.

I'm not sure how we could accommodate this issue without a significant re-work of how Helm handles subchart dependencies.

**Comment by kfox1111** [comment]
Values has supported subchart of subchart all along. That values does it, but conditionals use values, but not respect nesting is odd/unexpected.

I was involved in the design of conditionals early on and I don't remember any discussions on this design choice.

I think the assumption that postgresql should not be included twice is also not good. I could very well try to install say, keycloak and say, jupyterhub in one overarching chart, and want two separate databases, one per sub component.

It also comes up with some very weird api, where you may have to specify:
chart2.postgresql.enabled = true
and postgresql.enabled = true as chart2's templates can refer to its postgresql.enabled (but namespeced under chart2) (for rbac for example)

If you set the one but forget the other, it may cause some very strange things.

Its also impossible for the subchart with the conditional to know all names it can be nested under, as a dependency can be named when included in requirements like so:
```
  - name: chart2
    alias: mythingy
```

then it would be mythingy.postgresql.enabled but there'd be no way for a generic chart2 to ever be written/shared, which is the point of nested charts.

Sharing a db is a good feature, but I'm afraid this particular implementation causes more confusion and other issues then it solves for the one feature.

I think this needs to be reconsidered.

**Comment by CosminSontu** [comment]
Running on following version of helm:
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}

In my case i have the following situation:
common-chart 
intermediate-chart --(requires[dependencies.common-chart.enabled=false])--> common-chart
umbrella-chart --(requires[dependencies.common-chart.enabled=true])--> common-chart
umbrella-chart --(requires[dependencies.intermediate-chart.enabled=true])--> intermediate-chart

I noticed that because in umbrella chart i have [dependencies.common-chart.enabled = true],
it overrides the one in intermediate-chart, causing common-chart to appear twice in --dry-run output.

i managed to prevent this by renaming [dependencies.common-chart.enabled] to [umbrella-chart-dependencies.common-chart.enabled] to avoid collision.

**Comment by jascott1** [comment]
This is by far the biggest problem with the subchart system (I implemented it). Regardless of pros/cons it seems unintuitive for users to explicitly specify `requirements.yaml` conditions from the root of the values tree and prevents a default relative evaluation that many seem to expect. 

Imo the mitigation for v2 is more examples in the docs. At this point I think a change here is a breaking change and we should concentrate our energy on improving the experience in v3. Perhaps we should start a "subchart use case" mega thread?

**Comment by kfox1111** [comment]
depends on how far out v3 is I think. If a couple months, may not be worth fixing. Much more then that, we may keep getting people tripping over it.

It could be solved in v2 with some kind of flag that specifies behaviour so the old behaviour can be kept.

Yeah, having some kind of tracking of the issue so it can be fixed for sure in v3 would be good regardless of if it is to be solved in v2 or not.

**Comment by jascott1** [comment]
Agree we can solve in v2 if someone has the time and it doesn't break. We could add syntax to the condition values for special behavior so we don't have to carry another flag. Of course that might just lead to more confusion.

**Comment by bacongobbler** [comment]
We should definitely have a closer look at the requirement subsystem for Helm 3. We've already started some work by moving the requirements.yaml into the Chart.yaml as a `dependencies` field (see [the Helm 3 docs on this](https://github.com/splisson/helm/blob/25adb16388877fcb77227a1dc34d75902430dcc3/docs/charts.md#managing-dependencies-with-the-dependencies-field)), but if we're going to consider re-working the aliases, tags and condition fields, let's start thinking about it now and discuss potential solutions.

I'd prefer that we only do this for Helm 3, though. I like your suggestion about adding more examples in the docs as a mitigation strategy though, @jascott1. Most of the active Helm contributors are only looking at Helm 3 PRs unless it's a high prioity security issue or a major blocker/bugfix, so even if someone were to re-work the requirements subsystem for Helm 2 it might be ignored.

**Comment by kfox1111** [comment]
+1. I've seen some weirdness with aliases too.

**Comment by groundnuty** [comment]
@jascott1 I'm not 100% sure what you meant by 'subchart use case" mega thread' but I'm happy to provide a use case, as I've been [living with this problem for last 2 years](https://github.com/onedata/k8s-deployment/blame/7f3acbf39cbc9c0c93451a35bc94141d13c2fae7/landscapes/onedata-3p/landscape.yaml#L562). The [charts](https://github.com/onedata/charts) I work with have all the nice bits including: aliases (<<>> on the picture), conditional requirements, and 4 level dependency tree: 
![image](https://user-images.githubusercontent.com/241318/56503864-04c23380-6517-11e9-80d6-9b209eaaa0b9.png)

**Comment by bacongobbler** [comment]
this should now be fixed via https://github.com/helm/helm/pull/4917. https://github.com/helm/helm/issues/3734 raised the same issue/use case. Closing.
